server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfileBreakdown.java

```diff
@@ -46,12 +65,177 @@ public AbstractProfileBreakdown<QueryTimingType> context(Object context) {
     public Map<String, Long> toBreakdownMap() {
         final Map<String, Long> map = new HashMap<>(super.toBreakdownMap());
 
-        for (final AbstractProfileBreakdown<QueryTimingType> context : contexts.values()) {
-            for (final Map.Entry<String, Long> entry : context.toBreakdownMap().entrySet()) {
-                map.merge(entry.getKey(), entry.getValue(), Long::sum);
+        for (final Map.Entry<Object, AbstractProfileBreakdown<QueryTimingType>> context : contexts.entrySet()) {
+            Map<String, Long> breakdown = context.getValue().toBreakdownMap();
+            String collector = leafToCollector.get(context.getKey());
+            for (final Map.Entry<String, Long> entry : breakdown.entrySet()) {
+                String breakdownType = entry.getKey();
+                Long breakdownValue = entry.getValue();
+                // Adding breakdown type count
+                if (breakdownType.contains(TIMING_TYPE_COUNT_SUFFIX)) {
+                    map.merge(breakdownType, breakdownValue, Long::sum);
+                }
+                // Adding slice level breakdown
+                if (!breakdownType.contains(TIMING_TYPE_START_TIME_SUFFIX)) {
+                    String sliceLevelBreakdownType = breakdownType + AT + collector;
+                    map.put(sliceLevelBreakdownType, map.getOrDefault(sliceLevelBreakdownType, 0L) + breakdownValue);
+                }
             }
+            addMaxEndTimeAndMinStartTime(map, breakdown);
+            // Adding slice level latest leaf end time
+            long currentLeafEndTime = 0L;
+            for (QueryTimingType queryTimingType : QueryTimingType.values()) {
+                String timingType = queryTimingType.toString();
+                String timingTypeStartTime = timingType + AbstractProfileBreakdown.TIMING_TYPE_START_TIME_SUFFIX;
+                currentLeafEndTime = Math.max(currentLeafEndTime, breakdown.get(timingTypeStartTime) + breakdown.get(timingType));
+            }
+            String latestLeafEndTime = LATEST_LEAF_END_TIME + AT + collector;
+            map.put(latestLeafEndTime, Math.max(map.getOrDefault(latestLeafEndTime, 0L), currentLeafEndTime));
         }
+        return buildQueryProfileBreakdownMap(map);
```

Comment:
This is very hard to read. How about:
Use above `Map<Collector, List<LeafReaderContext>>` to iterate over all the entries. Then have 2 Map to keep track of slice level breakdowns and a global breakdowns separately. While iterating we can populate these 2 maps using each leaf breakdownMap entries. That will structure the code better IMO ?

Reply:
I have refactored the code. Could you please review it once more? Thanks.